crates/pessimistic-proof-core/src/local_state/commitment.rs

```diff
@@ -0,0 +1,103 @@
+//! State related commitments involved in the pessimistic proof.
+//!
+//! The pessimistic proof has the "pessimistic root" as part of its public
+//! inputs. Some logic in this file handles the migration on its computation.
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    bridge_exit::NetworkId,
+    keccak::{digest::Digest, keccak256_combine},
+    ProofError,
+};
+
+/// The state commitment of one [`super::NetworkState`].
+#[derive(Default, Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct StateCommitment {
+    pub exit_root: Digest,
+    pub balance_root: Digest,
+    pub nullifier_root: Digest,
+}
+
+/// The parameters which compose the pessimistic root.
+#[derive(Default, Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct PessimisticRoot {
+    pub balance_root: Digest,
+    pub nullifier_root: Digest,
+    pub ler_leaf_count: u32,
+    pub height: u64,
+    pub origin_network: NetworkId,
+}
+
+impl PessimisticRoot {
+    /// Infer the version of the provided pessimistic root.
+    pub fn infer_pp_root_version(&self, pp_root: Digest) -> Result<PPRootVersion, ProofError> {
+        // NOTE: Return v2 to trigger the migration
+        if pp_root.0 == [0u8; 32] {
+            return Ok(PPRootVersion::V2);
+        }
```

Comment:
We must assert here that the fields/roots aare emtpy, otherwise any "inital" value is psosible, that menas that:
```
 if self.balance_root != EMPTY_LER
                || self.nullifier_root != EMPTY_LER
                || self.ler_leaf_count != 0
                || self.height != 0 {
                return Err(ProofError::Unknown(
                    "Empty pp_root provided but internal fields are non-zero".to_string(),
                ));
            }
````

Reply:
good catch 👍

Reply:
actually not entirely fixed, I need to give the proper last pp root in `apply_certificate()` especially for the tests

Reply:
done